File: module/plugins/eltdetail/htdocs/js/graphs.js
/* Copyright (C) 2009-2015:
   Gabes Jean, naparuba@gmail.com
   Gerhard Lausser, Gerhard.Lausser@consol.de
   Gregory Starck, g.starck@gmail.com
   Hartmut Goebel, h.goebel@goebel-consult.de
   Andreas Karfusehr, andreas@karfusehr.de
   Frederic Mohier, frederic.mohier@gmail.com

   This file is part of Shinken.

   Shinken is free software: you can redistribute it and/or modify
   it under the terms of the GNU Affero General Public License as published by
   the Free Software Foundation, either version 3 of the License, or
   (at your option) any later version.

   Shinken is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU Affero General Public License for more details.

   You should have received a copy of the GNU Affero General Public License
   along with Shinken.  If not, see <http://www.gnu.org/licenses/>.
*/


function get_range() {
    g_s = graphstart;  // Time start and ends
    g_e = graphend;

    g_p = 0; // relative pos for start and end
    g_q = 0;

    // theses offsets are for PNP, with won't work with
    // Graphite from now.
    // TODO
    g_A = 65;
    g_B = 25;
    g_T = 587;

    g_D = g_T - (g_A + g_B); // The pixel range when we remove borders

    g_O = g_e - g_s; // total time printed
}

/*
   { }  <-- User selection.

         s    p             q  e         # s = time start, p = rel pos start, q = rel pos end, e = end time
         |    {             }  |
              x             x2           # x = select start, x2=select end
   [  A  ][       D           ][  B  ]   # A = offset left, D=usefull, B = offset right
   [              T                  ]   # T = Total size of the picture

*/

function update_coords(c)
{
    // variables can be accessed here as
    // c.x, c.y, c.x2, c.y2, c.w, c.h
    // Compute relative positions
    g_p = Math.min(Math.max(g_A, c.x), g_T - g_B) - g_A;
    g_q = Math.min(Math.max(g_A, c.x2), g_T - g_B) - g_A;
};


// We will compute the relative position of the selection
// by removing the borders. This will give us g_rp and g_rq.
// Then we compute the ratio of this selection, and so we apply
// it in the time selection. And we are done.
function graph_zoom(uri){
   //console.log(uri);

   //alert('Relatives'+g_p+' '+g_q);

   // We compute the ratio of the relative position from the inner
   // draw (without the borders)
   var g_rp = g_p / g_D;
   var g_rq = g_q / g_D;

   //alert('Relative ratio: '+g_D+' '+g_rp+' '+g_rq);

   // Now compute the new start and new end we want to border
   var g_ns = parseInt(g_s + g_O*g_rp);
   var g_ne = parseInt(g_s + g_O*g_rq);

   //alert('New time '+g_ns+' '+g_ne);

   // Maybe we just fuck up, if so, bailout
   if(g_ne <= g_ns){
      return;
   }

   // Make the uri and GO!
   var new_uri = uri+'graphstart='+g_ns+'&graphend='+g_ne+'#graphs';
   console.log(new_uri);
   window.location=new_uri;
}

File: module/plugins/graphs/graphs.py
#!/usr/bin/python

# -*- coding: utf-8 -*-

# Copyright (C) 2009-2012:
#    Gabes Jean, naparuba@gmail.com
#    Gerhard Lausser, Gerhard.Lausser@consol.de
#    Gregory Starck, g.starck@gmail.com
#    Hartmut Goebel, h.goebel@goebel-consult.de
#
# This file is part of Shinken.
#
# Shinken is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# Shinken is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.
#
# You should have received a copy of the GNU Affero General Public License
# along with Shinken.  If not, see <http://www.gnu.org/licenses/>.
import time
import json
import requests
import random
from shinken.log import logger


### Will be populated by the UI with it's own value
app = None


def proxy_graph():
    ''' This route proxies graphs returned by the graph module.
        The pnp4nagios/graphite image url have to be in the GET attributes,
        encoded with urlencode. The graphs metamodule takes care of that. This
        route should not be usefull anywhere else.
    '''
    url = app.request.GET.get('url', '')
    try:
        r = requests.get(url)
        if r.status_code != 200:
            logger.error("[WebUI-graph] Image URL not found: %d - %s", r.status_code, url)
            app.bottle.response.status = r.status_code
            app.bottle.response.content_type = 'application/json'
            return json.dumps(
                {'status': 'ko', 'message': r.content}
            )

    except Exception as e:
        logger.error("[WebUI-graph] exception: %s", str(e))
        app.bottle.response.status = 409
        app.bottle.response.content_type = 'application/json'
        return json.dumps(
            {'status': 'ko', 'message': str(e)}
        )

    app.bottle.response.content_type = str(r.headers['content-type'])
    return r.content


# Our page
def get_graphs_widget():
    user = app.request.environ['USER']
    # Graph URL may be: http://192.168.0.42/render/?width=320&height=240&fontSize=8&lineMode=connected&from=04:57_20151203&until=04:57_20151204&tz=Europe/Paris&title=Outlook_Web_Access/ - rta&target=alias(color(Outlook_Web_Access.rta,"green"),"rta")&target=alias(color(constantLine(1000),"orange"),"Warning")&target=alias(color(constantLine(3000),"red"),"Critical")
    url = app.request.GET.get('url', '')
    logger.debug("[WebUI-graph] graph URL: %s", url)

    if not url:
        search = app.request.GET.get('search', '') or app.datamgr.get_hosts(user)[0].host_name
        elt = app.datamgr.get_element(search, user) or app.redirect(404)
    else:
        search = app.request.GET.get('search', '')
        elt = None

    duration = app.request.GET.get('duration', '86400')
    duration_list = {
        '3600': '1h',
        '86400': '1d',
        '172800': '2d',
        '604800': '7d',
        '2592000': '30d',
        '31536000': '365d'
    }

    wid = app.request.query.get('wid', 'widget_graphs_' + str(int(time.time())))
    collapsed = (app.request.query.get('collapsed', 'False') == 'True')

    options = {
        'search': {
            'value': search,
            'type': 'hst_srv',
            'label': 'Element name'
        },
        'url': {
            'value': url,
            'type': 'text',
            'label': 'Graph URL'
        },
        'duration': {
            'value': duration,
            'values':  duration_list,
            'type': 'select',
            'label': 'Duration'
        },
    }

    title = 'Element graphs'
    if search:
        title = 'Element graphs for %s (%s)' % (search, duration_list[str(duration)])

    graphsId = "graphs_%d" % random.randint(1, 9999)

    return {
        'elt': elt,
        'wid': wid,
        'graphsId': graphsId,
        'collapsed': collapsed,
        'options': options,
        'base_url': '/widget/graphs',
        'url': url,
        'title': title,
        'duration': int(duration),
    }

widget_desc = '''<h4>Graphs</h4>
Show the perfdata graph
'''

pages = {
    proxy_graph: {
        'name': 'Graph', 'route': '/graph', 'view': 'graph', 'static': True
    },
    get_graphs_widget: {
        'name': 'wid_Graph', 'route': '/widget/graphs', 'view': 'widget_graphs', 'static': True, 'widget': ['dashboard'], 'widget_desc': widget_desc, 'widget_name': 'graphs', 'widget_picture': '/static/graphs/img/widget_graphs.png'
    }
}
